lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -14,60 +14,137 @@
+    ( Trace, logDebug, logNotice )
+import Cardano.Pool.DB
+    ( DBLayer (..), ErrPointAlreadyExists )
+import Cardano.Wallet.Network
+    ( ErrNetworkTip
+    , ErrNetworkUnavailable
+    , FollowAction (..)
+    , NetworkLayer (networkTip, stakeDistribution)
+    , follow
+    , staticBlockchainParameters
+    )
 import Cardano.Wallet.Primitive.Types
-    ( BlockHeader (..), EpochNo (..), PoolId (..), SlotId (..) )
+    ( BlockHeader (..), PoolId (..), SlotId (..) )
+import Control.Monad
+    ( forM_, when )
 import Control.Monad.IO.Class
-    ( MonadIO )
-import Control.Retry
-    ( RetryPolicyM, retrying )
-import Data.Either
-    ( isLeft )
+    ( liftIO )
+import Control.Monad.Trans.Class
+    ( lift )
+import Control.Monad.Trans.Except
+    ( runExceptT, throwE, withExceptT )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.List.NonEmpty
+    ( NonEmpty )
 import Data.Map.Merge.Strict
-    ( WhenMatched, WhenMissing, mergeA, traverseMissing, zipWithMatched )
+    ( mergeA, traverseMissing, zipWithMatched )
 import Data.Map.Strict
     ( Map )
 import Data.Quantity
     ( Quantity (..) )
+import Data.Text
+    ( Text )
 import Data.Word
     ( Word64 )
-import Fmt
-    ( Buildable (..), blockListF', fmt, (+|), (|+) )
 import GHC.Generics
     ( Generic )
 import Numeric.Natural
     ( Natural )
 
 import qualified Data.Map.Strict as Map
 
--- | For a given epoch, and state, this function returns /how many/ blocks
--- each pool produced.
-activityForEpoch :: EpochNo -> State -> Map PoolId Int
-activityForEpoch epoch s =
-    Map.filter (> 0)
-    $ Map.map (length . filter slotInCurrentEpoch)
-    (activity s)
-  where
-    slotInCurrentEpoch = ((epoch ==) . epochNumber)
+data Block = Block
+    { header :: BlockHeader
+    , producer :: PoolId
+    } deriving (Eq, Show, Generic)
 
-data ErrMetricsInconsistency
-    = ErrMetricsInconsistencyBlockProducerNotInStakeDistr
-        PoolId
-        (Quantity "block" Natural)
+-- | 'monitorStakePools' follows the chain and puts pool productions and stake
+-- distributions to a 'DBLayer', such that the data in the database is always
+-- consistent.
+--
+-- The pool productions and stake distrubtions in the db can /never/ be from
+-- different forks such that it's safe for readers to access it.
+--
+-- FIXME: This last statement is only true if 'putStakeProduction' and
+-- 'putPoolProduction' are running in the same db transaction, which would be
+-- the case if we make 'SqlPersistT' the top-level monad for the 'DBLayer'
+-- instead of 'IO'.
+monitorStakePools
+    :: NetworkLayer IO t Block
+    -> DBLayer IO
+    -> Trace IO Text
+    -> IO ()
+monitorStakePools nl db tr = do
+    cursor <- initCursor
+    follow nl tr cursor forward backward header
+  where
+    initCursor :: IO [BlockHeader]
+    initCursor = do
+        let (_block0, bp) = staticBlockchainParameters nl
+        let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+        readCursor db k
```

Comment:
Actually, something to be careful with: the cursor might be empty. In which case we should use the genesis hash as a starting point.